From 8d2664db1c1967848356b9082ce0e7b0796d3e61 Mon Sep 17 00:00:00 2001
From: Pawel MOLL <pawel.moll@st.com>
Date: Thu, 9 Apr 2009 11:35:31 +0100
Subject: [PATCH] sh_stm: Fixed 7105 INTC2 configuration

7105 INTC2 configuration was subtly wrong in group 0 definitions -
I2S2SPDIF0 interrupt is single bit group member, while the rest of
I2S2SPDIF ones is mated with the first SATA interrupt.

Signed-off-by: Pawel Moll <pawel.moll@st.com>
---
 arch/sh/kernel/cpu/sh4/setup-stx7105.c |   52 ++++++++++++++++++-------------
 1 files changed, 30 insertions(+), 22 deletions(-)

diff --git a/arch/sh/kernel/cpu/sh4/setup-stx7105.c b/arch/sh/kernel/cpu/sh4/setup-stx7105.c
index beadb4f..5845646 100644
--- a/arch/sh/kernel/cpu/sh4/setup-stx7105.c
+++ b/arch/sh/kernel/cpu/sh4/setup-stx7105.c
@@ -1344,11 +1344,13 @@ enum {
 	WDT,
 	HUDI,
 
-	I2S2SPDIF0, I2S2SPDIF1, I2S2SPDIF2, I2S2SPDIF3,	/* Group 0 */
-	SATA_DMAC,
+	I2S2SPDIF0,					/* Group 0 */
+	I2S2SPDIF1, I2S2SPDIF2, I2S2SPDIF3, SATA_DMAC,
 	SATA_HOSTC, DVP, STANDALONE_PIO, AUX_VDP_END_PROC,
 	AUX_VDP_FIFO_EMPTY, COMPO_CAP_BF, COMPO_CAP_TF,
-	PIO0, PIO1, PIO2,
+	PIO0,
+	PIO1,
+	PIO2,
 
 	PIO6, PIO5, PIO4, PIO3,				/* Group 1 */
 	SSC3, SSC2, SSC1, SSC0,				/* Group 2 */
@@ -1370,8 +1372,8 @@ enum {
 	ICAM3_KTE, ICAM3, KEY_SCANNER, MES,		/* Group 16 */
 
 	/* interrupt groups */
-	TMU2, RTC,
-	GROUP0_0, GROUP0_1, GROUP0_2,
+	TMU2,
+	GROUP0_0, GROUP0_1,
 	GROUP1, GROUP2, GROUP3,
 	GROUP4, GROUP5, GROUP6, GROUP7,
 	GROUP8, GROUP9, GROUP10, GROUP11,
@@ -1385,14 +1387,15 @@ static struct intc_vect vectors[] = {
 	INTC_VECT(WDT, 0x560),
 	INTC_VECT(HUDI, 0x600),
 
-	INTC_VECT(I2S2SPDIF0, 0xa00), INTC_VECT(I2S2SPDIF1, 0xa20),
-	INTC_VECT(I2S2SPDIF2, 0xa40), INTC_VECT(I2S2SPDIF3, 0xa60),
-	INTC_VECT(SATA_DMAC, 0xa80),
+	INTC_VECT(I2S2SPDIF0, 0xa00),
+	INTC_VECT(I2S2SPDIF1, 0xa20), INTC_VECT(I2S2SPDIF2, 0xa40),
+	INTC_VECT(I2S2SPDIF3, 0xa60), INTC_VECT(SATA_DMAC, 0xa80),
 	INTC_VECT(SATA_HOSTC, 0xb00), INTC_VECT(DVP, 0xb20),
 	INTC_VECT(STANDALONE_PIO, 0xb40), INTC_VECT(AUX_VDP_END_PROC, 0xb60),
 	INTC_VECT(AUX_VDP_FIFO_EMPTY, 0xb80), INTC_VECT(COMPO_CAP_BF, 0xba0),
 	INTC_VECT(COMPO_CAP_TF, 0xbc0),
-	INTC_VECT(PIO0, 0xc00), INTC_VECT(PIO1, 0xc80),
+	INTC_VECT(PIO0, 0xc00),
+	INTC_VECT(PIO1, 0xc80),
 	INTC_VECT(PIO2, 0xd00),
 
 	INTC_VECT(PIO6, 0x1000), INTC_VECT(PIO5, 0x1020),
@@ -1432,11 +1435,12 @@ static struct intc_vect vectors[] = {
 static struct intc_group groups[] = {
 	INTC_GROUP(TMU2, TMU2_TUNI, TMU2_TICPI),
 
-	INTC_GROUP(GROUP0_0, I2S2SPDIF0, I2S2SPDIF1, I2S2SPDIF2, I2S2SPDIF3),
-	INTC_GROUP(GROUP0_1, SATA_DMAC, SATA_HOSTC, DVP, STANDALONE_PIO),
-	INTC_GROUP(GROUP0_2, AUX_VDP_END_PROC, AUX_VDP_FIFO_EMPTY,
-		   COMPO_CAP_BF, COMPO_CAP_TF),
-	/* PIO0, PIO1, PIO2 are not part of any group */
+	/* I2S2SPDIF0 is a single bit group */
+	INTC_GROUP(GROUP0_0, I2S2SPDIF1, I2S2SPDIF2, I2S2SPDIF3, SATA_DMAC),
+	INTC_GROUP(GROUP0_1, SATA_HOSTC, DVP, STANDALONE_PIO,
+			AUX_VDP_END_PROC, AUX_VDP_FIFO_EMPTY,
+			COMPO_CAP_BF, COMPO_CAP_TF),
+	/* PIO0, PIO1, PIO2 are single bit groups */
 
 	INTC_GROUP(GROUP1, PIO6, PIO5, PIO4, PIO3),
 	INTC_GROUP(GROUP2, SSC3, SSC2, SSC1, SSC0),
@@ -1467,14 +1471,18 @@ static struct intc_prio_reg prio_registers[] = {
 	{ 0xffd00008, 0, 16, 4, /* IPRB */     {  WDT,    0,    0,     0 } },
 	{ 0xffd0000c, 0, 16, 4, /* IPRC */     {    0,    0,    0,  HUDI } },
 	{ 0xffd00010, 0, 16, 4, /* IPRD */     { IRL0, IRL1,  IRL2, IRL3 } },
-						/* 31-28,   27-24,   23-20,   19-16 */
-						/* 15-12,    11-8,     7-4,     3-0 */
-	{ 0x00000300, 0, 32, 4, /* INTPRI00 */ {       0,       0,    PIO2,    PIO1,
-						    PIO0, GROUP0_2, GROUP0_1, GROUP0_0 } },
-	{ 0x00000304, 0, 32, 4, /* INTPRI04 */ {  GROUP8,  GROUP7,  GROUP6,  GROUP5,
-						  GROUP4,  GROUP3,  GROUP2,  GROUP1 } },
-	{ 0x00000308, 0, 32, 4, /* INTPRI08 */ { GROUP16, GROUP15, GROUP14, GROUP13,
-						 GROUP12, GROUP11, GROUP10,  GROUP9 } },
+
+				/* 31-28,    27-24,    23-20,      19-16, */
+				/* 15-12,     11-8,      7-4,        3-0  */
+	{ 0x00000300, 0, 32, 4,
+		/* INTPRI00 */ {       0,        0,     PIO2,       PIO1,
+				    PIO0, GROUP0_1, GROUP0_0, I2S2SPDIF0 } },
+	{ 0x00000304, 0, 32, 4,
+		/* INTPRI04 */ {  GROUP8,   GROUP7,   GROUP6,     GROUP5,
+				  GROUP4,   GROUP3,   GROUP2,     GROUP1 } },
+	{ 0x00000308, 0, 32, 4,
+		/* INTPRI08 */ { GROUP16,  GROUP15,  GROUP14,    GROUP13,
+				 GROUP12,  GROUP11,  GROUP10,     GROUP9 } },
 };
 
 static struct intc_mask_reg mask_registers[] = {
-- 
1.6.0.6

